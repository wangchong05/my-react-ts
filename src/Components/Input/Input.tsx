import React, { useState, ReactNode } from "react";
import styled, { StyledProps } from "styled-components";

export interface InputProps extends StyledProps<any> {
  label?: string;
  type?: "text" | "date" | "password" | "email" | "radio";
  value?: string | number;
  readonly?: boolean;
  disabled?: boolean;
}

// const InputWrapper: React.FC<InputProps> = styled.div<InputProps>`
//   padding: 10px;
//   border: 1px solid gray;
// `;

const Input = (props: InputProps) => {
  const { 
    label, 
    KeyName, 
    type, 
    defaultValue, 
    readOnly, 
    changeContent, 
    disabled 
  }: InputProps = props;
  console.log({...props})

  const [inputType, setInputType] = useState(type)
  console.log(inputType)
  const [hide, setHide] = useState(true)

  const hidePassword = (): boolean => {
    const hideTemp = !hide;
    setHide(hideTemp)
    setInputType(hideTemp?"password":"text")
    return true
  }

  const onChangeContent = (e:any) => {
    // console.log(e.target.value)
    changeContent({[KeyName]: type === "password" ? window.btoa(e.target.value) : e.target.value})
  }

  const labelStyle = props.labelStyle || {
    'min-width': '100px',
    'display': 'inline-block',
    'color': 'black',
    'font-size': '16px'
  }
  // style样式的处理
  const styleFun = (style: any) => {
      let styleStr = '';
      const labelStyleKeys = Object.keys(style);
      labelStyleKeys.forEach((e: string) => {
        const labelStyleKeyName: any = style[e];
        styleStr = styleStr + e + ':' + labelStyleKeyName + ';'
      })
      console.log(styleStr)
      return styleStr;
  }

  const Span = styled.span`
    ${
      styleFun(labelStyle)
    }
  `;


  return (
    <div>
      <Span>{label}</Span>
      <input 
        // {...props}
        type={inputType} 
        defaultValue={defaultValue} 
        readOnly={readOnly} 
        disabled={disabled} 
        onChange={onChangeContent}/>
      { type === "password" && (
        <input type="button" value={hide? "show": "hide"} onClick={hidePassword}/>
      )}
    </div>
  );
}

Input.defaultProps = {
  type: 'text',
  value: '',
  readonly: false
};

export default Input;