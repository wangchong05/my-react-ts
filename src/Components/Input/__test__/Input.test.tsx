import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Input from '../Input';
 
describe('Input render', () => {
  test('renders visible static elements in the Input', () => {
    const { getByTestId } = render(<Input label={'first name'}></Input>)
    // screen.getByRole('')
    
    expect(getByTestId('input')).toBeInTheDocument();
    expect(getByTestId('span')).toHaveTextContent("first name");// label render
  });
});
 
describe('Props', () => {
  test('type is password, B should be shown', () => {
    const { getByTestId } = render(<Input type={'password'}></Input>)
    expect(getByTestId('input-button')).toBeInTheDocument();

    expect(getByTestId('labe-input')).toMatchSnapshot();
  });
  test('no type', () => {
    render(<Input></Input>);
    userEvent.type(screen.getByTestId('input'), 'input value');
    expect(screen.getByDisplayValue('input value')).toBeInTheDocument();
    expect(screen.getByTestId('input')).toHaveAttribute('type', 'text');
    
    expect(screen.getByTestId('labe-input')).toMatchSnapshot();
  });
  test.each([
    ['text', 'input text value', 'input text value'],
    ['password', 'input password value', 'input password value']
  ])('type is password or text', (type, val, expected) => {
    render(<Input type={type}></Input>);
    userEvent.type(screen.getByTestId('input'), val);
    expect(screen.getByDisplayValue(expected)).toBeInTheDocument();
    expect(screen.getByTestId('input')).toHaveAttribute('type', type);
    
    expect(screen.getByTestId('labe-input')).toMatchSnapshot();
  });
  test('type is date', () => {
    render(<Input type={'date'}></Input>);
    userEvent.type(screen.getByTestId('input'), '2017-04-01');
    expect(screen.getByDisplayValue('2017-04-01')).toBeInTheDocument();
    expect(screen.getByTestId('input')).toHaveAttribute('type', 'date');
    
    expect(screen.getByTestId('labe-input')).toMatchSnapshot();
  });

  test('incoming defaultValue, the textbox should show the same value', async () => {
    const { getByTestId } = render(<Input defaultValue={'Tom'}></Input>);
    expect(getByTestId('input')).toHaveValue("Tom");
    
    expect(getByTestId('labe-input')).toMatchSnapshot();
  });

  test('readOnly, the textbox should have readOnly attribute', async () => {
    const { getByTestId } = render(<Input readOnly></Input>);
    expect(getByTestId('input')).toHaveAttribute('readOnly');
    
    expect(getByTestId('labe-input')).toMatchSnapshot();
  });

  test('not readOnly, the textbox should not have readOnly attribute', async () => {
    const { getByTestId } = render(<Input></Input>);
    expect(getByTestId('input')).not.toHaveAttribute('readOnly');
    
    expect(getByTestId('labe-input')).toMatchSnapshot();
  });

  test('disabled', async () => {
    const { getByTestId } = render(<Input disabled></Input>);
    expect(getByTestId('input')).toHaveAttribute('disabled')
    
    expect(getByTestId('labe-input')).toMatchSnapshot();
  });

  test('not disabled', async () => {
    const { getByTestId } = render(<Input></Input>);
    expect(getByTestId('input')).not.toHaveAttribute('disabled');
    
    expect(getByTestId('labe-input')).toMatchSnapshot();
  });

  test('style', async () => {
    const { getByTestId } = render(<Input labelStyle={{'min-width': '100px','font-size': '14px'}}></Input>);
    expect(getByTestId('span')).toHaveStyle({'min-width': '100px','font-size': '14px'});
    
    expect(getByTestId('labe-input')).toMatchSnapshot();
  });

  test('type is password, when the Show Password button is clicked, the correct text can be displayed', async () => {
    const { getByTestId } = render(<Input type={'password'}></Input>);
    expect(getByTestId('input-button')).toHaveValue("show");
    await userEvent.click(getByTestId('input-button'));
    expect(getByTestId('input-button')).toHaveValue("hide");
    
    expect(getByTestId('labe-input')).toMatchSnapshot();
  });
});
