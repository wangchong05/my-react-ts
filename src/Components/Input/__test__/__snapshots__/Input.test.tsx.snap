// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Props disabled 1`] = `
<div
  data-testid="labe-input"
  style="display: flex; margin: 10px;"
>
  <span
    class="sc-pNWdM iDMmfe"
    data-testid="span"
  />
  <input
    data-testid="input"
    disabled=""
    style="min-width: 200px; border: 1px solid gray;"
    type="text"
    value=""
  />
</div>
`;

exports[`Props incoming defaultValue, the textbox should show the same value 1`] = `
<div
  data-testid="labe-input"
  style="display: flex; margin: 10px;"
>
  <span
    class="sc-gKAaRy ezFlSV"
    data-testid="span"
  />
  <input
    data-testid="input"
    style="min-width: 200px; border: 1px solid gray;"
    type="text"
    value="Tom"
  />
</div>
`;

exports[`Props no type 1`] = `
<div
  data-testid="labe-input"
  style="display: flex; margin: 10px;"
>
  <span
    class="sc-dlnjwi lcIRfM"
    data-testid="span"
  />
  <input
    data-testid="input"
    style="min-width: 200px; border: 1px solid gray;"
    type="text"
    value=""
  />
</div>
`;

exports[`Props not disabled 1`] = `
<div
  data-testid="labe-input"
  style="display: flex; margin: 10px;"
>
  <span
    class="sc-jrsJWt hKxyeA"
    data-testid="span"
  />
  <input
    data-testid="input"
    style="min-width: 200px; border: 1px solid gray;"
    type="text"
    value=""
  />
</div>
`;

exports[`Props not readOnly, the textbox should not have readOnly attribute 1`] = `
<div
  data-testid="labe-input"
  style="display: flex; margin: 10px;"
>
  <span
    class="sc-fujyAs xZBSA"
    data-testid="span"
  />
  <input
    data-testid="input"
    style="min-width: 200px; border: 1px solid gray;"
    type="text"
    value=""
  />
</div>
`;

exports[`Props readOnly, the textbox should have readOnly attribute 1`] = `
<div
  data-testid="labe-input"
  style="display: flex; margin: 10px;"
>
  <span
    class="sc-iCoGMd fKbeHV"
    data-testid="span"
  />
  <input
    data-testid="input"
    readonly=""
    style="min-width: 200px; border: 1px solid gray;"
    type="text"
    value=""
  />
</div>
`;

exports[`Props style 1`] = `
<div
  data-testid="labe-input"
  style="display: flex; margin: 10px;"
>
  <span
    class="sc-kEqXSa fhsrSR"
    data-testid="span"
  />
  <input
    data-testid="input"
    style="min-width: 200px; border: 1px solid gray;"
    type="text"
    value=""
  />
</div>
`;

exports[`Props type is date 1`] = `
<div
  data-testid="labe-input"
  style="display: flex; margin: 10px;"
>
  <span
    class="sc-jSFjdj liZikR"
    data-testid="span"
  />
  <input
    data-testid="input"
    style="min-width: 200px; border: 1px solid gray;"
    type="date"
    value=""
  />
</div>
`;

exports[`Props type is password or text 1`] = `
<div
  data-testid="labe-input"
  style="display: flex; margin: 10px;"
>
  <span
    class="sc-hKFxyN ctvA-DE"
    data-testid="span"
  />
  <input
    data-testid="input"
    style="min-width: 200px; border: 1px solid gray;"
    type="text"
    value=""
  />
</div>
`;

exports[`Props type is password or text 2`] = `
<div
  data-testid="labe-input"
  style="display: flex; margin: 10px;"
>
  <span
    class="sc-eCApnc bIgqzE"
    data-testid="span"
  />
  <input
    data-testid="input"
    style="min-width: 200px; border: 1px solid gray;"
    type="password"
    value=""
  />
  <input
    data-testid="input-button"
    type="button"
    value="show"
  />
</div>
`;

exports[`Props type is password, B should be shown 1`] = `
<div
  data-testid="labe-input"
  style="display: flex; margin: 10px;"
>
  <span
    class="sc-gtsrHT hfZTVk"
    data-testid="span"
  />
  <input
    data-testid="input"
    style="min-width: 200px; border: 1px solid gray;"
    type="password"
    value=""
  />
  <input
    data-testid="input-button"
    type="button"
    value="show"
  />
</div>
`;

exports[`Props type is password, when the Show Password button is clicked, the correct text can be displayed 1`] = `
<div
  data-testid="labe-input"
  style="display: flex; margin: 10px;"
>
  <span
    class="sc-crzoAE gFAeUq"
    data-testid="span"
  />
  <input
    data-testid="input"
    style="min-width: 200px; border: 1px solid gray;"
    type="text"
    value=""
  />
  <input
    data-testid="input-button"
    type="button"
    value="hide"
  />
</div>
`;
